drivers/w5500/w5500.c

```diff
@@ -0,0 +1,470 @@
+    return spi_transfer_byte(dev->p.spi, dev->p.cs, false, 0u);
+}
+
+static void write_register(w5500_t *dev, uint16_t reg, uint8_t data)
+{
+    uint16_t address = reg & 0x07ff;
+    uint8_t command = (uint8_t)((reg & 0xf800) >> 8u) | CMD_WRITE;
+
+    send_addr(dev, address);
+    spi_transfer_byte(dev->p.spi, dev->p.cs, true, command);
+    spi_transfer_byte(dev->p.spi, dev->p.cs, false, data);
+}
+
+static uint16_t read_addr(w5500_t *dev, uint16_t addr_high, uint16_t addr_low)
+{
+    uint16_t res = (read_register(dev, addr_high) << 8);
+
+    res |= read_register(dev, addr_low);
+    return res;
+}
+
+static void write_address(w5500_t *dev,
+                  uint16_t addr_high, uint16_t addr_low, uint16_t val)
+{
+    write_register(dev, addr_high, (uint8_t)(val >> 8));
+    write_register(dev, addr_low, (uint8_t)(val & 0xff));
+}
+
+static void read_chunk(w5500_t *dev, uint16_t addr, uint8_t *data, size_t len)
+{
+    uint16_t address = addr & 0x07ff;
+    uint8_t command = (uint8_t)((addr & 0xf800) >> 8u) | CMD_READ;
+
+    send_addr(dev, address);
+    spi_transfer_byte(dev->p.spi, dev->p.cs, true, command);
+    spi_transfer_bytes(dev->p.spi, dev->p.cs, false, NULL, data, len);
+}
+
+static void write_chunk(w5500_t *dev, uint16_t addr, uint8_t *data, size_t len)
+{
+    uint16_t address = addr & 0x07ff;
+    uint8_t command = (uint8_t)((addr & 0xf800) >> 8u) | CMD_WRITE;
+
+    send_addr(dev, address);
+    spi_transfer_byte(dev->p.spi, dev->p.cs, true, command);
+    spi_transfer_bytes(dev->p.spi, dev->p.cs, false, data, NULL, len);
+}
+
+static void extint(void *arg)
+{
+    w5500_t *dev = (w5500_t *)arg;
+
+    netdev_trigger_event_isr(&dev->netdev);
+    if (dev->p.polling_interval_ms > 0u) {
+        /* restart timer if we are polling */
+        (void)ztimer_set(ZTIMER_MSEC, &dev->timerInstance, dev->p.polling_interval_ms);
+    }
+}
+
+void w5500_setup(w5500_t *dev, const w5500_params_t *params, uint8_t index)
+{
+    assert(dev);
+    assert(params);
+
+    /* Initialize netdev structure. */
+    dev->netdev.driver = &netdev_driver_w5500;
+    dev->netdev.event_callback = NULL;
+    dev->netdev.context = dev;
+    /* Initialize the device descriptor. */
+    dev->p = *params;
+
+    dev->link_up = false;
+
+    /* Initialize the chip select pin. */
+    spi_init_cs(dev->p.spi, dev->p.cs);
+    if (gpio_is_valid(dev->p.evt)) {
+        /* Initialize the external interrupt pin. */
+        gpio_init_int(dev->p.evt, GPIO_IN, GPIO_FALLING, extint, dev);
+    }
+    netdev_register(&dev->netdev, NETDEV_W5500, index);
+}
+
+static bool check_configuration(w5500_t *dev)
+{
+    /* Polling mode: W5500_PARAM_EVT == GPIO_UNDEF and polling_interval_ms > 0u    */
+    /* Interrupt mode: W5500_PARAM_EVT != GPIO_UNDEF and polling_interval_ms == 0u */
+    return((dev->p.evt == GPIO_UNDEF && dev->p.polling_interval_ms > 0u) ||
+           (dev->p.evt != GPIO_UNDEF && dev->p.polling_interval_ms == 0u));
+}
```

Comment:
I'd be a bit more graceful here: You could have a valid polling interval always configured by default, but when a IRQ pin is provided use that instead of polling (and instead of producing an unnecessary error on init)

Reply:
Ok. My intention was to point the user to the invalid configuration, but I have softened the edge now :-)